nonffi/writer.go

```diff
@@ -0,0 +1,191 @@
+package nonffi
+
+import (
+	"math/bits"
+	"runtime"
+	"sync"
+
+	"github.com/filecoin-project/go-commp-utils/zerocomm"
+	"github.com/filecoin-project/go-fil-commcid"
+	"github.com/filecoin-project/go-fil-commp-hashhash"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+type DataCIDSize struct {
+	PayloadSize int64
+	PieceSize   abi.PaddedPieceSize
+	PieceCID    cid.Cid
+}
+
+const commPBufPad = abi.PaddedPieceSize(8 << 20)
+const CommPBuf = abi.UnpaddedPieceSize(commPBufPad - (commPBufPad / 128)) // can't use .Unpadded() for const
+
+type ciderr struct {
+	c   cid.Cid
+	err error
+}
+
+type Writer struct {
+	len    int64
+	buf    [CommPBuf]byte
+	leaves []chan ciderr
+
+	tbufs    [][CommPBuf]byte
+	throttle chan int
+}
+
+func (w *Writer) Write(p []byte) (int, error) {
+	if w.throttle == nil {
+		w.throttle = make(chan int, runtime.NumCPU())
+		for i := 0; i < cap(w.throttle); i++ {
+			w.throttle <- i
+		}
+	}
+	if w.tbufs == nil {
+		w.tbufs = make([][CommPBuf]byte, cap(w.throttle))
+	}
+
+	n := len(p)
+	for len(p) > 0 {
+		buffered := int(w.len % int64(len(w.buf)))
+		toBuffer := len(w.buf) - buffered
+		if toBuffer > len(p) {
+			toBuffer = len(p)
+		}
+
+		copied := copy(w.buf[buffered:], p[:toBuffer])
```

Comment:
Data copies is one of the most expensive things you can do. Could you not iterate over subslices of the buffer direcly handed to `Write()` above?

Basically this set of steps: https://github.com/filecoin-project/go-fil-commp-hashhash/blob/d66fe63391de/commp.go#L197-L217

Reply:
Those buffers are used in background goroutines to compute stuff in parallel in between calls to write - the thread where the Write is called is only doing copying, which is actually not that expensive compared to the rest of CpmmP

Reply:
(I believe that slices passed into Write can’t be used once the method returns)